### DISCUSSION TITLE
[PSD4] Esercizio Slide 1

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/psd2223#136] [Esercizi (A-L)]
**Esercizio (1+2+1 punti)**

(a) Rappresentare X = - 97 e Y = 39 in Ca2, ognuno con il minimo numero di bit.

(b) Dopo aver calcolato il numero di bit necessario per rappresentare sia la somma X+Y che la differenza X-Y, portare X e Y alla lunghezza necessaria ed eseguire le due operazioni.

(c) Infine, verificare i risultati ottenuti.

**Comment by FeddyLix17** [comment]
**(a)**
- $\large X = -97_{10} = 10011111_{Ca2}$ [bit usati: 8]
- $\large Y = 39_{10} = 100111_{Ca2}$ [bit usati: 6]

**(b)**
- Somma:
  $\large X+Y = -97_{10} + 39_{10} = 10011111_{Ca2} + 100111_{Ca2} = 11000110_{Ca2} (-58_{10})$ [bit usati: 8]

- Differenza:
 $\large X - Y = -97_{10} - 39_{10} = -97_{10} - (+39_{10}) = -97_{10} + (-39_{10}) =$
 $\large = 10011111_{Ca2} + 11011001_{Ca2} = 101111000_{Ca2} (-136_{10})$ [bit usati: 9]

- $\large X = -97_{10} = 110011111_{Ca2}$ [bit usati: 9]
- $\large Y = 39_{10} = 00100111_{Ca2}$ [bit usati: 8]
- $\large Y = 39_{10} = 000100111_{Ca2}$ [bit usati: 9]
- $\large -39_{10} = 11011001_{Ca2}$ [bit usati: 8]

**(c)**
- i vari risultati dovrebbero essere corretti

**Comment by matypist** [comment]
**(a)**

- $X = -97_{10}$ viene rappresentato come $10011111$ in Ca2 con $8$ bit.
- $Y = 39_{10}$ viene rappresentato come $100111$ in Ca2 con $6$ bit.

**(b)**

- Somma $X + Y$:

  - Il valore assoluto massimo tra $X = -97$ e $Y = 39$ è $97$, rappresentabile in Ca2 con $8$ bit per come avevamo determinato. Aggiungendo un bit extra per il bit di carry, otteniamo un totale di $8 + 1 = 9$ bit massimi necessari per la rappresentazione della somma.

  - $10011111$ +
    $00100111$ =
    \--------------
    $11000110$

  - La somma $X + Y = -97 + 39 = -58$ viene rappresentata come $11000110$ in Ca2 con $8$ bit.

- Differenza $X - Y$:

  - Per quanto riguarda il numero di bit necessario si hanno considerazioni analoghe a quelle per la somma, visto che la differenza viene calcolata come $X + (-Y)$, corrispondendo ad una somma e richiedendo dunque $8 + 1 = 9$ bit.
  
  - Per poter effettuare la somma dovremo quindi innanzitutto calcolare il $Ca_2$ di $Y$ a $8$ bit, ottenendo $-Y = -39_{10} = 11011001_{Ca2}$

  -     $10011111$ +
        $11011001$ =
   \----------------
    $101111000$

  - La differenza $X - Y = -97 - 39 = -136$ viene rappresentata come $101111000$ in Ca2 con $9$ bit.

**(c)**

- I risultati ottenuti sono corretti e corrispondono alle operazioni eseguite, la verifica della loro correttezza può corrispondere ad esempio a convertire in decimale i due risultati $11000110$ e $101111000$, dalla quale conversione ci accorgiamo facilmente che corrispondono a quanto avremmo ottenuto calcolando direttamente $X + Y$ e $X - Y$ in decimale, ovvero rispettivamente $-58_{10}$ e $-136_{10}$.

**Comment by FeddyLix17** [comment]
@matypist ho corretto la piccola svista sul minimo dei bit necessari a rappresentare +39 inizialmente, comunque trovo la tua soluzione più esaustiva. In base alla reazione che mi hai dato volevo sapere se avessi trovato qualche altro errore nella mia soluzione.

**Comment by matypist** [comment]
> @matypist ho corretto la piccola svista sul minimo dei bit necessari a rappresentare +39 inizialmente, comunque trovo la tua soluzione più esaustiva. In base alla reazione che mi hai dato volevo sapere se avessi trovato qualche altro errore nella mia soluzione.

Hey @FedVlogger17, non mi ero nemmeno accorto di quella svista in realtà xD, diciamo che l'uso dell'emoji "confused" come reazione era relativo al fatto che il tutto mi sembrava scritto un po' confusionario, o comunque non mi sembrava assolvere la, seppur stranissima, richiesta da parte della consegna di calcolare **in precedenza** il numero di bit necessario per rappresentare somma e differenza (che in realtà però non saprei assicurarti che io stesso abbia ben assolto)

I risultati comunque sembrano coincidere tra le nostre due soluzioni, e per questo ho etichettato l'esercizio come risolto.
